@tool
extends Node2D
#class_name SpinningObstacle

@export var spiningCenterTexture: Texture
@export_range(0.01, 0.5, 0.05) var spiningCenterTextureScale: float = 0.1

@export_group("Stick Properties")
@export var stickTexture: Texture
@export_range(0.1, 1.0, 0.05) var stickTextureScale: float = 1.0
@export var rotated: bool = false
@export_range(-1000, 1000, 50, "or_less", "or_greater") var rotatingSpeed: float = 100.0

@export_group("Pendulum Properties")
@export var pendulumTexture: Texture
@export var animatedPendulumSprite: SpriteFrames
@export var animeName: StringName = "default"
@export_range(0.1, 1.0, 0.05) var pendulumScale: float = 1.0
@export var collisionShape: Shape2D

@export_group("Pendulum 2 Properties")
@export var pendulum2Texture: Texture
@export var animatedPendulum2Sprite: SpriteFrames
@export var anime2Name: StringName = "default"
@export_range(0.1, 1.0, 0.05) var pendulum2Scale: float = 1.0
@export var collisionShape2: Shape2D

@export_group("Others")
@export var player: CharacterBody2D
@export_range(0, 100, 5, "or_greater") var damageValue: int = 25

@export var visiblityNotifier: bool = true


@onready var spining_center: Marker2D = $SpiningCenter
@onready var stick_sprite: Sprite2D = $SpiningCenter/StickSprite
@onready var spining_center_sprite: Sprite2D = $SpiningCenter/SpiningCenterSprite

@onready var pendulum: Area2D = $SpiningCenter/Pendulum
@onready var collision_shape: CollisionShape2D = $SpiningCenter/Pendulum/CollisionShape2D
@onready var pendulum_sprite: Sprite2D = $SpiningCenter/Pendulum/PendulumSprite
@onready var animated_pendulum_sprite: AnimatedSprite2D = $SpiningCenter/Pendulum/AnimatedPendulumSprite

@onready var pendulum_2: Area2D = $SpiningCenter/Pendulum2
@onready var collision_shape_2: CollisionShape2D = $SpiningCenter/Pendulum2/CollisionShape2D2
@onready var pendulum_2_sprite: Sprite2D = $SpiningCenter/Pendulum2/Pendulum2Sprite
@onready var animated_pendulum_2_sprite: AnimatedSprite2D = $SpiningCenter/Pendulum2/AnimatedPendulum2Sprite

@onready var visible_on_screen_notifier: VisibleOnScreenNotifier2D = $VisibleOnScreenNotifier2D


func _ready() -> void:
	apply_properties()

func _process(_delta: float) -> void:
	if Engine.is_editor_hint():
		apply_properties()

func _physics_process(delta: float) -> void:
	spining_center.rotation_degrees += rotatingSpeed * delta


func apply_properties() -> void:
	if spiningCenterTexture:
		spining_center_sprite.texture = spiningCenterTexture
		spining_center_sprite.scale.x = spiningCenterTextureScale
		spining_center_sprite.scale.y = spiningCenterTextureScale
	
	if stickTexture:
		stick_sprite.texture = stickTexture
		stick_sprite.scale.x = stickTextureScale
		stick_sprite.scale.y = stickTextureScale
		
		if rotated:
			stick_sprite.rotation_degrees = 90.0
		else:
			stick_sprite.rotation_degrees = 0.0
		
		#stick_sprite.position.y = (stick_sprite.texture.get_height() * stick_sprite.scale.y) / 2
	
	if animatedPendulumSprite:
		pendulum_sprite.texture = null
		animated_pendulum_sprite.sprite_frames = animatedPendulumSprite
		animated_pendulum_sprite.scale.x = pendulumScale
		animated_pendulum_sprite.scale.y = pendulumScale
		
		if !animated_pendulum_sprite.is_playing():
			animated_pendulum_sprite.play(animeName)
		
		pendulum.position.y = (stick_sprite.texture.get_height() * stick_sprite.scale.y) / 2
	
	elif pendulumTexture:
		animated_pendulum_sprite.stop()
		pendulum_sprite.texture = pendulumTexture
		pendulum_sprite.scale.x = pendulumScale
		pendulum_sprite.scale.y = pendulumScale
		
		pendulum.position.y = (stick_sprite.texture.get_height() * stick_sprite.scale.y) / 2
	
	
	if animatedPendulum2Sprite:
		pendulum_2_sprite.texture = null
		animated_pendulum_2_sprite.sprite_frames = animatedPendulum2Sprite
		animated_pendulum_2_sprite.scale.x = pendulum2Scale
		animated_pendulum_2_sprite.scale.y = pendulum2Scale
		
		if !animated_pendulum_2_sprite.is_playing():
			animated_pendulum_2_sprite.play(anime2Name)
		
		pendulum_2.position.y = -(stick_sprite.texture.get_height() * stick_sprite.scale.y) / 2
	
	elif pendulum2Texture:
		animated_pendulum_2_sprite.stop()
		pendulum_2_sprite.texture = pendulum2Texture
		pendulum_2_sprite.scale.x = pendulum2Scale
		pendulum_2_sprite.scale.y = pendulum2Scale
		
		pendulum_2.position.y = -(stick_sprite.texture.get_height() * stick_sprite.scale.y) / 2
	
	
	if collisionShape:
		collision_shape.shape = collisionShape
	
	if collisionShape2:
		collision_shape_2.shape = collisionShape2
	
	visible_on_screen_notifier.visible = visiblityNotifier

func _on_pendulum_body_entered(body: Node2D) -> void:
	if body == player:
		body.health -= damageValue
		body.runtime_vars.damaged = true
		body.camera.shake(0.2, Vector2(2.0, 2.0))

func _on_pendulum_2_body_entered(body: Node2D) -> void:
	if body == player:
		body.health -= damageValue
		body.runtime_vars.damaged = true
		body.camera.shake(0.2, Vector2(2.0, 2.0))
